channeldb/invoices.go

```diff
@@ -412,6 +412,195 @@ func (d *DB) GetInvoiceWithTx(ctx context.Context, dbTx database.DBTx,
+		invoiceIndexBucket,
+	)
+	if invoiceIndex == nil {
+		return invpkg.ErrNoInvoicesCreated
+	}
+
+	invoiceAddIndex := invoices.NestedReadWriteBucket(
+		addIndexBucket,
+	)
+	if invoiceAddIndex == nil {
+		return invpkg.ErrNoInvoicesCreated
+	}
+
+	// settleIndex can be nil, as the bucket is created lazily when the
+	// first invoice is settled.
+	settleIndex := invoices.NestedReadWriteBucket(settleIndexBucket)
+	payAddrIndex := tx.ReadWriteBucket(payAddrIndexBucket)
+
+	for _, ref := range refs {
+		invoice, err := d.GetInvoiceWithTx(ctx, tx, ref)
+		if err != nil {
+			return invpkg.ErrInvoiceNotFound
+		}
+
+		// Fetch the invoice key for using it to check for consistency
+		// and also to delete from the invoice index.
+		var payHash lntypes.Hash
+		switch {
+		case ref.PayHash() != nil:
+			payHash = *ref.PayHash()
+
+		case invoice.Terms.PaymentPreimage != nil:
+			payHash = invoice.Terms.PaymentPreimage.Hash()
+
+		default:
+			return errors.New("unable to find payment hash or " +
+				"preimage")
+		}
+
+		invoiceKey := invoiceIndex.Get(payHash[:])
+		if invoiceKey == nil {
+			return invpkg.ErrInvoiceNotFound
+		}
+
+		err = invoiceIndex.Delete(payHash[:])
+		if err != nil {
+			return err
+		}
+
+		// Delete payment address index reference if there's a valid
+		// payment address passed.
+		payAddr := invoice.Terms.PaymentAddr
+		if payAddr == invpkg.BlankPayAddr {
+			// To ensure consistency check that the already fetched
+			// invoice key matches the one in the payment address
+			// index.
+			key := payAddrIndex.Get(payAddr[:])
+			if bytes.Equal(key, invoiceKey) {
+				// Delete from the payment address index. Note
+				// Note that since the payment address index
+				// has been introduced with an empty migration
+				// it may be possible that the index doesn't
+				// have an entry for this invoice.
+				// ref: https://github.com/lightningnetwork/lnd/pull/4285/commits/cbf71b5452fa1d3036a43309e490787c5f7f08dc#r426368127
+				if err := payAddrIndex.Delete(
+					payAddr[:],
+				); err != nil {
+					return err
+				}
+			}
+		}
+
+		var addIndexKey [8]byte
+		byteOrder.PutUint64(addIndexKey[:], invoice.AddIndex)
+
+		// To ensure consistency check that the key stored in the add
+		// index also matches the previously fetched invoice key.
+		key := invoiceAddIndex.Get(addIndexKey[:])
+		if !bytes.Equal(key, invoiceKey) {
+			return fmt.Errorf("unknown invoice " +
+				"in add index")
+		}
+
+		// Remove from the add index.
+		err = invoiceAddIndex.Delete(addIndexKey[:])
+		if err != nil {
+			return err
+		}
+
+		// Remove from the settle index if available and if the invoice
+		// is settled.
+		if invoice.SettleIndex != 0 {
+			var settleIndexKey [8]byte
+			byteOrder.PutUint64(
+				settleIndexKey[:], invoice.SettleIndex,
+			)
+
+			// To ensure consistency check that the already fetched
+			// invoice key matches the one in the settle index.
+			key := settleIndex.Get(settleIndexKey[:])
```

Comment:
not here. Before it could be nil because i was a point to a ref. Now we are using the value of an `uint64`, which will always have a value. The value 0 is considered "nil" or not-set yet